src/Nethermind/Nethermind.Blockchain/Processing/BlockchainProcessor.cs

```diff
@@ -377,6 +317,58 @@ public Block Process(Block suggestedBlock, ProcessingOptions options, IBlockTrac
             return lastProcessed;
         }
 
+        private Block[] ProcessBranch(
+            ProcessingOptions options,
+            IBlockTracer tracer,
+            ProcessingBranch processingBranch)
+        {
+            Block[] processedBlocks = null;
+            try
+            {
+                processedBlocks = _blockProcessor.Process(
+                    processingBranch.Root,
+                    processingBranch.BlocksToProcess,
+                    options,
+                    tracer);
+            }
+            catch (InvalidBlockException ex)
+            {
+                // only run it for the first processing
+                if ((options & ProcessingOptions.RerunWithTraceOnFailure)
+                    == ProcessingOptions.RerunWithTraceOnFailure)
+                {
+                    GethLikeBlockTracer gethTracer
+                        = new GethLikeBlockTracer(GethTraceOptions.Default);
+                    ProcessBranch(
```

Comment:
Recursion here introduces confusion, can we just extract a method for this?

Reply:
let me see